torchgen/gen.py

```diff
@@ -2175,9 +2307,30 @@ def gen_op_headers(
             ),
         }
 
+    all_groups: List[
+        Union[NativeFunction, NativeFunctionsGroup, NativeFunctionsViewGroup]
+    ] = list(structured_native_functions) + list(
+        view_groups  # type: ignore[assignment, arg-type, operator]
+    )
+    # Note: all operators that functionalization needs to handle (mutable and aliasing ops) should be grouped properly.
+    # The only reason we really need to deal with direct NativeFunctions here (instead of the groups) is because:
+    # (1) We can provide better error checking (error out if someone introduces a mutable op that doesn't obey the grouping logic)
+    # (2) functionalization needs to manually register CompositeImplicitAutograd kernels, which might not be grouped.
+    #     Although this could go away long-term if we add a dedicated dispatch key for decompositions.
+    structured_map: Dict[OperatorName, NativeFunction] = {
+        f.func.name: f
+        for f in concatMap(lambda g: list(g.functions()), structured_native_functions)
+    }
+    view_map: Dict[OperatorName, NativeFunction] = {
+        f.func.name: f for f in concatMap(lambda g: list(g.functions()), view_groups)
+    }
+    for f in native_functions:
+        if f.func.name not in structured_map and f.func.name not in view_map:
+            all_groups.append(f)
```

Comment:
So with this PR, you will only end up with an un-grouped NativeFunction if:
- it's `CompositeImplicitAutograd` (because we don't bother generating the other variants)
- it has a functional schema, and no other variants exist (we could totally generate an out= variant for these, but this PR doesn't currently)
- it has the `manual_cpp_binding` flag (we probably don't care about grouping these, since they don't go through the dispatcher and therefore don't have any codegen logic associated with them?)
- it's in one of the hardcoded "problematic" lists that I added, due to having a bad schema

So what I mean by that is you can't really "accidentally" add a new NativeFunction that doesn't get grouped. Either if fulfills the criteria to get generated variants (that let us group it), or the codegen explicitly doesn't bother generating groups for it.

I do want to add better error messages around whether or not you expected the codegen to generate variants for you (see the comments here in model.py: https://github.com/pytorch/pytorch/pull/76320/files#r871474773)

Reply:
Ho so we cannot add more "problematic" functions by mistake?

Reply:
yep :)

Ok well technically:
- You can no longer add an op that is problematic the way the `_cummax/min` are (we'll trip an error saying you have a `SchemaKind.mutable` op that we can't gen an out= variant for
- Right now with this PR, you can potentially add a functional + inplace op manually to the yaml that you intended to be grouped, but don't get grouped if you messed up their schemas. We probably want the codegen to error in that case (instead of generating a brand new functional variant to pair up with your inplace variant). So I'm going to add another error check. Probably something like:

```
- func: _fused_moving_avg_obs_fq_helper(...
  dispatch:
    CPU: fused_moving_avg_obs_fake_quant_cpu
    CUDA: fused_moving_avg_obs_fake_quant_cuda
  # the autogen tag doesn't actually tell the codegen what to generate; it's just used for error checking.
  # codegen will figure out to automatically generate a functional + out= variant based on this op's schema,
  # and then later assert that this `autogen` label was present on the original op
  autogen: _fused_moving_avg_obs_fq_helper.functional, _fused_moving_avg_obs_fq_helper.out
```